### PR TITLE
[Clang] [Sema] Reject non-power-of-2 `_BitInt` matrix element types

### PR DESCRIPTION
**Pull request by Sirraide**
Essentially, this pr makes this ill-formed:
```c++
using mat4 = _BitInt(12) [[clang::matrix_type(3, 3)]];
```

This matches preexisting behaviour for vector types (e.g. `ext_vector_type`), and given that LLVM IR intrinsics for matrices also take vector types, it seems like a sensible thing to do.

This is currently especially problematic since we sometimes lower matrix types to LLVM array types instead, and while e.g. `[4 x i32]` and `<4 x i32>` *probably* have the same similar memory layout (though I don’t think it’s sound to rely on that either, see #117486), `[4 x i12]` and `<4 x i12>` definitely don’t.

### CODE REVIEW

File: clang/lib/Sema/SemaType.cpp [hunk 1]
```diff
@@ -2312,6 +2312,18 @@ QualType Sema::BuildArrayType(QualType T, ArraySizeModifier ASM,
   return T;
 }
 
+bool CheckBitIntElementType(Sema &S, SourceLocation AttrLoc,
```
Comment (tbaederr) [inline clang/lib/Sema/SemaType.cpp:2315]:
```suggestion
static bool CheckBitIntElementType(Sema &S, SourceLocation AttrLoc,
```

File: clang/lib/Sema/SemaType.cpp [hunk 2]
```diff
@@ -2312,6 +2312,18 @@ QualType Sema::BuildArrayType(QualType T, ArraySizeModifier ASM,
...
+                            const BitIntType *BIT, bool ForMatrixType = false) {
+  // Only support _BitInt elements with byte-sized power of 2 NumBits.
+  unsigned NumBits = BIT->getNumBits();
+  if (!llvm::isPowerOf2_32(NumBits) || NumBits < 8) {
+    S.Diag(AttrLoc, diag::err_attribute_invalid_bitint_vector_type)
+        << ForMatrixType << (NumBits < 8);
+    return true;
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaType.cpp:2322]: AFAIK you can do
```suggestion
    return S.Diag(AttrLoc, diag::err_attribute_invalid_bitint_vector_type)
        << ForMatrixType << (NumBits < 8);
```

File: clang/lib/Sema/SemaType.cpp [hunk 3]
```diff
@@ -2455,6 +2456,10 @@ QualType Sema::BuildMatrixType(QualType ElementTy, Expr *NumRows, Expr *NumCols,
     return QualType();
   }
 
+  if (const auto *BIT = ElementTy->getAs<BitIntType>();
+      BIT && CheckBitIntElementType(*this, AttrLoc, BIT, true))
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaType.cpp:2460]:
```suggestion
      BIT && CheckBitIntElementType(*this, AttrLoc, BIT, /*ForMatrixType=*/true))
```

**Review by Fznamznon — COMMENTED** [review]
It feels like that the non-power-of-2 _BitInt vectors were disabled because it wasn't known how to handle them. https://reviews.llvm.org/D133634#3793653
It probably is still so, however, does it make sense to document the new behavior? 
Hopefully also, since this never worked properly, this is not a breaking change.

**Review by fhahn — COMMENTED** [review]
Makes sense to me, but do we have tests that make sure codegen for matrixes with power-of-2 bitints is sane?

File: clang/lib/Sema/SemaType.cpp (hunk 2 above)
Comment (Sirraide) [reply clang/lib/Sema/SemaType.cpp:2322]: Ah yeah, I always forget that you can do this.

**Comment by Sirraide** [comment]
Interestingly, it seems we don’t even have tests for `_BitInt` vectors...

**Comment by Sirraide** [comment]
Er, well this doesn’t look right:

```c++
using i8x3 = _BitInt(32) __attribute__((ext_vector_type(3)));
i8x3 v1(i8x3 a) { return a; }
```

```llvm
%a.addr = alloca <3 x i32>, align 16
%extractVec = shufflevector <3 x i32> %a, <3 x i32> poison, <4 x i32> <i32 0, i32 1, i32 2, i32 poison>
store <4 x i32> %extractVec, ptr %a.addr, align 16
```

We’re storing a `<4 x i32>` into a `<3 x i32>` alloca; I don’t think that’s sound, is it?

**Comment by efriedma-quic** [comment]
The type of an alloca is only used to determine the number of bytes allocated; it doesn't have any other semantics.  So it's fine as long as the alloca is 16 bytes.  Which I think it is in the given example, because of the datalayout rules for `<3 x i32>`.

**Comment by Sirraide** [comment]
> The type of an alloca is only used to determine the number of bytes allocated;

Yeah, that part I know; I was mainly just wondering if there’s a platform/abi/whatever where that doesn’t work out.

**Comment by efriedma-quic** [comment]
(For context, grep for PreserveVec3Type.)

Probably the code should be defensive and check for the allocation size of the vector type, but I don't think any targets define the alignment for non-power-of-2 vectors in a way that would cause issues.

**Comment by Sirraide** [comment]
> an integer type (as in C23 6.2.5p22)

Doesn’t 6.2.5p22 include 6.2.5.p5 i.e. `_BitInt`s?

CC @AaronBallman

**Comment by Sirraide** [comment]
I’ve added some tests and also updated the docs to mention that non-power-of-2 `_BitInt`s are not valid matrix element types.

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Pull request merged**